syntax = "proto3";

import "response.proto";

package YrcfAgent;

message BucketConfig {
    string host_name = 1;
    string bucket_name = 2;

    enum Protocol {
        HTTPS = 0;
        HTTP = 1;
    }
    Protocol protocol = 3;

    enum UriStyle {
        PATH = 0;
        VIRTUAL_HOST = 1;
    }
    UriStyle uri_style = 4;
    string region = 5;
    string access_key = 6;
    string secret_access_key = 7;

    enum LibraryType {
        LIBS3 = 0;
        AWS = 1;
    }
    LibraryType lib_type = 8;
    string token = 9;
}

message BucketInfo {
    uint32 bucket_id = 1;
    BucketConfig bucket_config = 2;
}

//message for list bucket
message ListBucketsRequest {}
message ListBucketsResponse {
    ResMes result = 1;
    repeated BucketInfo bucket_info = 2;
}

//message for add and test bucket
message AddOrTestBucketRequest {
    enum Operation {
        ADD_BUCKET = 0;
        TEST_BUCKET = 1;
    }
    Operation op = 1;
    BucketConfig bucket_config = 2;
}
message AddOrTestBucketResponse {
    ResMes result = 1;
}

//message for bucket update
message UpdateBucketRequest {
    uint32 bucket_id = 1;
    string access_key = 2;
    string secret_access_key = 3;
}
message UpdateBucketResponse {
    ResMes result = 1;
}

service Bucket {
    rpc ListBuckets(ListBucketsRequest) returns (stream ListBucketsResponse);
    rpc AddOrTestBucket(AddOrTestBucketRequest) returns (AddOrTestBucketResponse);
    rpc UpdateBucket(UpdateBucketRequest) returns (UpdateBucketResponse);
}